### DISCUSSION TITLE
Is there any plugin that can turn jsx code into code playground?

### DISCUSSION THREAD

**Discussion by aweiu** [mdx-js/mdx#1902] [Q&A] [answered]
Suppose the content of mdx is:

![image](https://user-images.githubusercontent.com/15623620/149887831-08874182-bfef-4835-b221-7b804b5c451a.png)

The result displayed by the browser is:

![image](https://user-images.githubusercontent.com/15623620/149887998-ca7cee2a-cc35-49fe-9d58-530f047ab12e.png)

Or a demo card like this:

![image](https://user-images.githubusercontent.com/15623620/149888304-3b6820c4-560f-460b-b27c-d49966541bf0.png)

**Comment by ChristianMurphy** [comment]
Are you looking for a playground for JSX codeblocks?
If so https://github.com/kevin940726/remark-codesandbox may be what you are looking for.

If you are looking for the source of the playground on the site, it isn't a plugin, but it's available at https://github.com/mdx-js/mdx/blob/main/docs/_component/editor.client.js

If you are looking for a more full featured documentation solution, there are several mdx based ones:
https://storybook.js.org/docs/react/api/mdx
https://docusaurus.io/docs/markdown-features/react

**Reply by aweiu** [reply]
Thanks for your reply! But none of those you listed is what I want.

My requirement is to "**convert all the JSX code of the user in an mdx file into playground**"

---

Now, I am writing a remark plug-in to implement it, But I have a big problem：**How can multiple JSX code blocks in the same mdx file isolate the scope?**，for example:

```
## Example 1

\```jsx
import TestComponent from './Test'

const testVariable = 1
\```

## Example 2

\```jsx
import TestComponent from './Test'

const testVariable = 2
\```
```

You will receive an error that `Testcomponent` and `testvariable` are redefined.

I don't know how to generate mdast in the remark plug-in to realize the scope isolation of each JSX block

**Reply by ChristianMurphy** [reply]
> How can multiple JSX code blocks in the same mdx file isolate the scope?

I don't quite understand, what do you want the scope isolated from?

---

In your example you didn't use codeblocks, did you mean to? (if so use quaduple backticks in the outer codeblock ` ```` ` so you can use triple back ticks in the block ` ``` `)
Could you also share an example of how you are trying to run the code? Maybe in a sandbox:

| description | codesandbox | stackblitz |
| - | - | - |
| MDX Loader with Next.js | [codesandbox](https://codesandbox.io/s/github/mdx-js/.github/tree/main/sandbox-templates/mdx-loader-next) | [stackblitz](https://stackblitz.com/github/mdx-js/.github/tree/main/sandbox-templates/mdx-loader-next) | 
| MDX Evaluate with Node.js | [codesandbox](https://codesandbox.io/s/github/mdx-js/.github/tree/main/sandbox-templates/mdx-evaulate-node) | [stackblitz](https://stackblitz.com/github/mdx-js/.github/tree/main/sandbox-templates/mdx-evaulate-node) |
| MDX Evaluate with Create React App | [codesandbox](https://codesandbox.io/s/github/mdx-js/.github/tree/main/sandbox-templates/mdx-evaluate-create-react-app) | [stackblitz](https://stackblitz.com/github/mdx-js/.github/tree/main/sandbox-templates/mdx-evaluate-create-react-app) |

**Reply by aweiu** [reply]
Coming! sorry to bother you again:

https://codesandbox.io/s/magical-river-0pkmm?file=/next.config.mjs

**Comment by wooorm** [comment, marked as answer]
A new thread, but responding to https://github.com/mdx-js/mdx/discussions/1902#discussioncomment-1995786.

The biggest problem with your approach: JSX is not MDX. If you want JSX, don’t use MDX. If you want MDX, don’t use ` ```jsx `. Use Babel or so when you want to evaluate JS(X).

Further problems with your code:

- You are using MDX in MDX. This is unrelated to what you want to do (turning code into a playground)
- The code for the editor component on our website [is here](https://github.com/mdx-js/mdx/blob/main/docs/_component/editor.client.js)
- The—hastily written—plugin used on our website to enhance certain code blocks in MDX files, [is here](https://github.com/mdx-js/mdx/blob/04c7faedf7131495410c55ff8fb37403483d8c08/website/mdx-config.js#L192)

**Reply by aweiu** [reply]
https://user-images.githubusercontent.com/15623620/150123331-d0380ce9-0f7d-4ab8-b834-503b32ddcff0.mp4

I have now implemented it by import a demo file instead of converting JSX directly.

After I finish it, I will share this remark plugin on GitHub.(Now it's too simple)
